File: client/lib/monarch/model/relations/selection.js
constructor("Model.Relations.Selection", Model.Relations.Relation, {

  initialize: function(operand, predicate) {
    this.operand = operand;
    this.predicate = predicate;
    this.initialize_events_system();
  },

  records: function() {
    if (this._records) return this._records;

    var predicate = this.predicate;
    var records = [];
    this.operand.each(function() {
      if (predicate.evaluate(this)) records.push(this);
    });
    return records;
  },

  wire_representation: function() {
    return {
      type: "selection",
      operand: this.operand.wire_representation(),
      predicate: this.predicate.wire_representation()
    };
  },

  subscribe_to_operands: function() {
    this.memoize_records();

    var self = this;
    this.operands_subscription_bundle.add(this.operand.on_insert(function(record) {
      if (self.predicate.evaluate(record)) self.record_inserted(record);
    }));

    this.operands_subscription_bundle.add(this.operand.on_remove(function(record) {
      if (self.predicate.evaluate(record)) self.record_removed(record);
    }));

    this.operands_subscription_bundle.add(this.operand.on_update(function(record, changed_fields) {
      if (self.contains(record)) {
        if (self.predicate.evaluate(record)) {
          self.record_updated(record, changed_fields);
        } else {
          self.record_removed(record);
        }
      } else {
        if (self.predicate.evaluate(record)) self.record_inserted(record);
      }
    }));
  },

  evaluate_in_repository: function(repository) {
    return new Model.Relations.Selection(this.operand.evaluate_in_repository(repository), this.predicate);
  },

  primary_table: function() {
    return this.operand.primary_table();
  },

  column: function(name) {
    return this.operand.column(name);
  }
});

File: server/lib/monarch/model/relations/selection.rb
module Model
  module Relations
    class Selection < Relation
      class << self
        def from_wire_representation(representation, repository)
          operand = Relation.from_wire_representation(representation["operand"], repository)
          predicate = Predicates::Predicate.from_wire_representation(representation["predicate"], repository)
          new(operand, predicate)
        end
      end

      attr_reader :operand, :predicate
      delegate :column, :surface_tables, :build_record_from_database, :to => :operand

      def initialize(operand, predicate, &block)
        super(&block)
        @operand, @predicate = operand, predicate
      end

      def create(field_values={})
        operand.create(predicate.force_matching_field_values(field_values))
      end

      def build_sql_query(query=Sql::Select.new)
        query.add_condition(predicate)
        operand.build_sql_query(query)
      end

      def ==(other)
        return false unless other.instance_of?(self.class)
        operand == other.operand && predicate == other.predicate
      end

      protected

      def subscribe_to_operands
        operand_subscriptions.add(operand.on_insert do |record|
          on_insert_node.publish(record) if predicate.matches?(record)
        end)

        operand_subscriptions.add(operand.on_update do |record, changeset|
          if predicate.matches?(changeset.old_state)
            if predicate.matches?(changeset.new_state)
              on_update_node.publish(record, changeset)
            else
              on_remove_node.publish(record)
            end
          else
            if predicate.matches?(changeset.new_state)
              on_insert_node.publish(record)
            end
          end
        end)

        operand_subscriptions.add(operand.on_remove do |record|
          on_remove_node.publish(record) if predicate.matches?(record)
        end)
      end
    end
  end
end
